### PR TITLE
docs: add page on types

### PR DESCRIPTION
**Pull request by Aflynn50**
Add a page on doc types and update other pages to make consistent.

### CODE REVIEW

File: docs/howto/prepare-and-run.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+(prepare-and-run)=
+# How to prepare and run a SQLair query
+
+A SQLair query is just like a regular SQL query, except that it contains special
+input and output expressions. These indicate to SQLair the Go objects that your
+inputs and outputs correspond to. To create and run a SQLair query there are
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md]: Move "To create..." to a new paragraph and cut out "there are several steps" -- just put a colon and then the steps.

File: docs/howto/prepare-and-run.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+3. Prepare your query with `sqlair.Prepare`.
+4. Run your query on a database or transaction.
+
+## Prepare your Go types
+
+See {ref}`types` for how to prepare your types.
+
+## Write your query
+
+SQLair queries are regular SQL queries with special input and output expressions
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md:18]: This repeats content from the intro. FWIW it feels more relevant here, so maybe you can skip it in the intro.

File: docs/howto/prepare-and-run.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+inserted which allow you to reference the Go types directly from the SQL. You
+can have any number of input and output expressions in your query.
+
+To write expressions that map Go types to query inputs see {ref}`input-expressions`.
+
+To write expressions that map Go types to query outputs see {ref}`output-expressions`.
+
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md]: Adopt a formulaic way of linking. E.g.,

> See more: {ref}`input-expressions`, {ref}`output-expressions`

The result is still not great but imo this is a move in the right direction.

File: docs/howto/prepare-and-run.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+If a query contains no output expressions then SQLair will not fetch any results
+from the database. This means that you will not get a `ErrNoRows` error if you
+run a `SELECT` statement with no output expressions to check if a row exists.
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md]: This sentence is hard to process. Try to make the story more linear.

File: docs/howto/prepare-and-run.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+You will need to include a placeholder output expression.
+
+## Prepare your query
+
+Now you have your query string, you need to pass it to `sqlair.Prepare` along
+with all the types mentioned in the input and output expression.
+
+See more [`sqlair.Preapre`](https://pkg.go.dev/github.com/canonical/sqlair#Prepare)
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md]: Use > in front of the See more. Helps separate at-issue content from meta content.

File: docs/howto/prepare-and-run.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+This is needed so that SQLair can gather reflection information about the types
+and verify the input/output expressions. If there is an issue with the actual
+SQL in the query this will not catch it.
+
+Note that this will not prepare the query on the database. In fact, the
+`sqlair.Statement` you get back here can be used on any database or transaction.
+The query is automatically prepared against the database the first time it is
+run. As a user, you do not need to worry about managing the driver prepared
+statements on the database. SQLair manages this for you.
+
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md:42]: This is a lot of words. Wondering how to tell the same story more efficiently.

File: docs/howto/prepare-and-run.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+## Run your query
+
+To run your query you will need a SQLair database
```
Comment (tmihoc) [inline docs/howto/prepare-and-run.md]: What does "you will need" mean? Do I have to do anything before I run those things in the next step? If not, this paragraph can be removed.

File: docs/howto/types.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
+(types)=
+# How to prepare your Go types 
+
```
Comment (tmihoc) [inline docs/howto/types.md:3]: Update the section titles so the structure of the HTG becomes clearer. Do you mean:

How to prepare your Go types for structs
How to prepare your Go types for maps
How to prepare your Go types for slices

?

PS "Prepare your Go types" is not a complete thought -- the story here is that you're preparing them for SQLair. That should be in all the titles. I'm also struggling to see where this prepare step fits in the overall story.

File: docs/howto/types.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+supports inputs from structs maps and slices and can output into structs
+and maps. 
+
+In general, all types used with SQLair need to be named, otherwise there is no
+way to reference them in the query. The rest of the preparation of the type
+depends on what kind it is.
+
+## Structs
+
+To use a struct with SQLair you must tag all the fields that correspond to a
```
Comment (tmihoc) [inline docs/howto/types.md]: This "must" -- is it trying to describe a prerequisite or rather a step? Are there other steps? If not, just say:

> To prepare a struct Go type for use with SQLair, tag all the fields that correspond...

Basically, "must" is describing a truth, but here you're trying to give steps, so adjust the perspective accordingly. PS Also relevant for further content below.

File: docs/howto/types.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+This struct and its corresponding columns can be referenced from a SQLair input
+or output expression.
+
+If a struct contains an embedded struct then SQLair will treat the fields of the
+embedded structs as if they were fields in the parent struct.
+
+## Maps
+
+Named maps with a base key type of string can be used with SQLair. The value
```
Comment (tmihoc) [inline docs/howto/types.md]: Again, try to switch from Reference language (description) to How-to guide language (commands).

**Review by tmihoc — APPROVED** [review]
Approving, but please the comments. Let's talk again tomorrow.

File: docs/howto/prepare-and-run.md (hunk 2 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md:18]: Good point

File: docs/howto/prepare-and-run.md (hunk 3 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md]: Changed to that

File: docs/howto/prepare-and-run.md (hunk 4 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md]: I think this actually belongs on the output expression page. Have moved it there

File: docs/howto/prepare-and-run.md (hunk 5 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md]: Good point, done that everywhere

File: docs/howto/prepare-and-run.md (hunk 6 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md:42]: One way to do that would be not to tell it here, this is not strictly relevant.

Comment (Aflynn50) [reply docs/howto/prepare-and-run.md:42]: I think I'll leave this here for now, but would like to find a new home for it in the future. There a bit more that can be said on the subject so I think putting it in here doesn't do it justice. Its quite a cool feature of SQLair

File: docs/howto/prepare-and-run.md (hunk 7 above)
Comment (Aflynn50) [reply docs/howto/prepare-and-run.md]: It means you need to have this object already created

File: docs/howto/types.md (hunk 8 above)
Comment (Aflynn50) [reply docs/howto/types.md:3]: Yea, we should have a discussion about how this fits in, its not complete in my head either. This information needs to be somewhere though

**Pull request merged**